packages/flutter/lib/src/material/text_field.dart

```diff
@@ -1067,9 +1067,23 @@ class _TextFieldState extends State<TextField> with RestorationMixin implements
     });
   }
 
+
+  bool _isOnlyAffinityChanged(TextSelection selection, TextSelection other) {
+    return selection.affinity != other.affinity &&
+        selection.copyWith(affinity: TextAffinity.upstream) == other.copyWith(affinity: TextAffinity.upstream);
+  }
+
+  TextSelection? _lastSelection;
   void _handleSelectionChanged(TextSelection selection, SelectionChangedCause? cause) {
+    bool onlyAffinityChanged = false;
+    if (_lastSelection != null) {
+      onlyAffinityChanged = _isOnlyAffinityChanged(_lastSelection!, selection);
+     }
+    _lastSelection = selection;
+
     final bool willShowSelectionHandles = _shouldShowSelectionHandles(cause);
-    if (willShowSelectionHandles != _showSelectionHandles) {
+    if (willShowSelectionHandles != _showSelectionHandles &&
+        !(onlyAffinityChanged && cause == SelectionChangedCause.keyboard)) {
```

Comment:
I know this a little hack.
 If we can figure out why the engine change the affinity maybe we have a better solution.
CC @LongCatIsLooong

Reply:
The engine doesn't know/care about text affinity I believe. It always sets it to downstream.

Reply:
I have correct the value when receive the remote value. WDYT?